src/protocol/libp2p/kademlia/store.rs

```diff
@@ -115,6 +122,101 @@ impl MemoryStore {
         }
     }
 
+    /// Try to get providers from local store for `key`.
+    ///
+    /// Returns a non-empty list of providers, if any.
+    pub fn get_providers(&mut self, key: &Key) -> Option<&Vec<ProviderRecord>> {
+        let drop = self.provider_keys.get_mut(key).map_or(false, |providers| {
+            let now = std::time::Instant::now();
+            providers.retain(|p| !p.is_expired(now));
+
+            providers.is_empty()
+        });
+
+        if drop {
+            self.provider_keys.remove(key);
+
+            None
+        } else {
+            self.provider_keys.get(key)
+        }
+    }
+
+    /// Try to add a provider for `key`. If there are already `max_providers_per_key` for
+    /// this `key`, the new provider is only inserted if its closer to `key` than
+    /// the furthest already inserted provider. The furthest provider is then discarded.
+    ///
+    /// Returns `true` if the provider was added, `false` otherwise.
+    pub fn put_provider(&mut self, provider_record: ProviderRecord) -> bool {
+        // Make sure we have no more than `max_provider_addresses`.
+        let provider_record = {
+            let mut record = provider_record;
+            record.addresses.truncate(self.config.max_provider_addresses);
+            record
+        };
+
+        let can_insert_new_key = self.provider_keys.len() < self.config.max_provider_keys;
+
+        match self.provider_keys.entry(provider_record.key.clone()) {
+            Entry::Vacant(entry) =>
+                if can_insert_new_key {
+                    entry.insert(vec![provider_record]);
+
+                    true
+                } else {
+                    tracing::warn!(
+                        target: LOG_TARGET,
+                        max_provider_keys = self.config.max_provider_keys,
+                        "discarding a provider record, because the provider key limit reached",
+                    );
+
+                    false
+                },
+            Entry::Occupied(mut entry) => {
+                let mut providers = entry.get_mut();
+
+                // Providers under every key are sorted by distance, with equal distances meaning
+                // peer IDs (more strictly, their hashes) are equal.
+                let provider_position =
+                    providers.binary_search_by(|p| p.distance().cmp(&provider_record.distance()));
+
+                match provider_position {
+                    Ok(i) => {
+                        // Update the provider in place.
+                        let existing_provider =
+                            providers.get_mut(i).expect("index was found in the vector; qed");
```

Comment:
nit: would `providers[i] = provider_record` work here?

Reply:
Probably yes, wanted to highlight that the index is correct. But this should be obvious.

Reply:
Done